### PR TITLE
1.1.15_wmagent wmstats replication temp fix

### PR DESCRIPTION
**Pull request by ticoann**
This patch requires WMAgent couchapp update as well.

### CODE REVIEW

**Comment by ticoann** [comment]
If this works we just need to run agent with this patch until cmsweb upgrade. then revert this patch and create new patch with #8855

File: src/python/WMComponent/AnalyticsDataCollector/AnalyticsPoller.py [hunk 1]
```diff
@@ -77,6 +78,12 @@ def setup(self, parameters):
             pluginFactory = WMFactory("plugins", "WMComponent.AnalyticsDataCollector.Plugins")
             self.plugin = pluginFactory.loadObject(classname=self.pluginName)
 
+        # delete all agent-request type doc.
+        self.localSummaryCouchDB.deleteAllAgentRequestDocument()
```
Comment (amaltaro) [inline src/python/WMComponent/AnalyticsDataCollector/AnalyticsPoller.py]: Seangchan, what happens to the job-related documents? Can they be recreated and wmstats repopulated with them?

File: src/python/WMCore/Services/WMStats/WMStatsWriter.py [hunk 2]
```diff
@@ -99,6 +99,38 @@ def bulkUpdateData(self, docs, existingDocs):
...
+                doc['_rev'] = "%s-%s" % (int(revList[0]) + 1, revList[1])
+                self.couchDB.queue(doc)
+                revCache[doc['_id']] = doc['_rev']
+            else:
+                if secondTry:
+                    doc['_rev'] = "1-123456789"
+                    self.couchDB.queue(doc)
+                    revCache[doc['_id']] = doc['_rev']
+                else:
+                    idParts = doc['_id'].split('-')
```
Comment (amaltaro) [inline src/python/WMCore/Services/WMStats/WMStatsWriter.py]: Shouldn't it be `doc['_rev']...`? And what happens if it's the first time we create the document, it doesn't have a revision number yet.

File: src/python/WMComponent/AnalyticsDataCollector/AnalyticsPoller.py [hunk 3]
```diff
@@ -77,6 +78,12 @@ def setup(self, parameters):
...
+    
+        # populate cache: This will take time ~5 min
+        self.revCache = self.centralWMStatsCouchDB.getAllAgentRequestFromCentralServer()
```
Comment (amaltaro) [inline src/python/WMComponent/AnalyticsDataCollector/AnalyticsPoller.py]: How about encapsulating it in a try/except and printing a nice error message if it fails (likely timeout) and assign an empty dict to `self.revCache`?

**Comment by ticoann** [comment]
This seems to be working in my vm. Except the deletion part. I deletes the document with different revision number. I am not sure how it get that number.

**Comment by ticoann** [comment]
@amaltaro, I changed repfilter not to replicate agent_request doc directly instead of rotating the database. So far it seems OK. If I don't see the couch server crash until tomorrow morning. I will update drained production agents tomorrow and one non draining production agent to test out.

**Comment by ticoann** [comment]
I don't see the replication errors anymore. I will apply this patch to all drained agent.
This patch requires WMAgent couchapp update.

**Comment by amaltaro** [comment]
awesome!

**Comment by ticoann** [comment]
Alan, I applied patch now to below.
draining agent (submit6, submit8, vocms0250, vocms0252)
active agent (vocms0280).
(twiki is updated)

I will monitor today and if everything is ok I will apply to all other agent and update trello board.

**Comment by ticoann** [comment]
Now all the agents are updated.

draining agent (submit6, submit8, vocms0250, vocms0252)
active agent (submit7,   vocms0256. vocms0280, vocms0281,  vocms0282).

fingers crossed. :-) I will merge this and create the new patch for when cmsweb is upgraded.

**Comment by amaltaro** [comment]
Thank you very much, Seangchan.

From the list of workflows with missing data in WMStats, provided by Sharad in this ticket https://github.com/dmwm/WMCore/issues/8851

the only one left without any information yet is:
task_EXO-RunIISummer16MiniAODv3-10105

Hopefully it's just a patch that wasn't properly applied to one agent. I'll try to check it this morning

**Comment by amaltaro** [comment]
This workflow
pdmvserv_task_EXO-RunIISummer16MiniAODv3-10105__v1_T_180923_230353_8407

was being worked on by vocms0250, which turns not to have been properly patched (we missed the src code part). It's done, services+agents restarted and couchapps pushed (again).

**Comment by amaltaro** [comment]
vocms0252 was also missing the src/ changes. I updated this patch over there and also double check all the other active agents and I can see their information in wmstats. So we should be good by now.

Sorry if I lost track, but how are we going to handle it in the coming agent release (master branch?) What we need to revert once CMSWEB is upgrade?

Thank you very much for fixing it, Seangchan.

**Comment by amaltaro** [comment]
[Fix #8859 - wmstats replication temp fix](https://trello.com/c/EMbY3SEx/131-fix-8859-wmstats-replication-temp-fix)

**Comment by ticoann** [comment]
Sorry Alan, was there any conflict on vocms0252 or couch app is updated? I remember I applied patch there. When cmsweb is upgrade we will remove the cache and use new view directly. Although the cache refresh is not so bad (took < 1min in most cases), I think we can make it consistent. I will create the patch.

**Comment by amaltaro** [comment]
I didn't check whether the couchapp was updated, only that the javascript code was updated, while the src code wasn't. No problem, it's all sorted now and we can start shutting agents down again (submit6 was the first one ;))

**Pull request merged**